### PR TITLE
SoundSources: prefer FFmpeg to MediaFoundation

### PR DESCRIPTION
**Pull request by Be-ing**
SoundSourceMediaFoundation has known issues and has not been well
maintained. FFmpeg is now available on Windows with the new vcpkg
build environment.
https://bugs.launchpad.net/mixxx/+bug/1899242

What are the implications for shifted cues and beatgrids?

### CODE REVIEW

**Comment by Be-ing** [comment]
It would be great to have this crossplatform wrapper library as the highest priority on all platforms. When we switch to vcpkg on macOS for 2.4, FFmpeg will be available on macOS too.

**Comment by daschuer** [comment]
The idea was not changing any sound sources in 2.3.
In 2.4 we can than starting to verify the offset using the silence markers.

**Comment by Be-ing** [comment]
Attempting to account for offsets with all possible combinations of files and codecs seems quixotic to me.

So what do you propose we do about https://bugs.launchpad.net/mixxx/+bug/1899242 for 2.3? Nothing?

**Comment by uklotzde** [comment]
If no Windows Dev is willing to take over SoundSourceMediaFoundation won't get fixed.

File: src/sources/soundsourceffmpeg.cpp [hunk 1]
```diff
@@ -454,11 +454,7 @@ QStringList SoundSourceProviderFFmpeg::getSupportedFileExtensions() const {
 SoundSourceProviderPriority SoundSourceProviderFFmpeg::getPriorityHint(
         const QString& supportedFileExtension) const {
     Q_UNUSED(supportedFileExtension)
-    // TODO: Increase priority to Default or even Higher for all
-    // supported and tested file extension?
-    // Currently it is only used as a fallback after all other
-    // SoundSources failed to open a file or are otherwise unavailable.
-    return SoundSourceProviderPriority::Lowest;
+    return SoundSourceProviderPriority::Higher;
```
Comment (uklotzde) [inline src/sources/soundsourceffmpeg.cpp:457]: This has already been rejected (limited to M4A/AAC) and would prioritize SoundSourceFFmpeg for MP3/M4A/AAC/ALAC/AIFF/WAV on all platforms.

This would also be ambiguous without raising the priority for OGG/FLAC/OPUS from `Higher` to `Highest` (which we should do anyway).

**Comment by Be-ing** [comment]
@uklotzde what are the practical implications of https://bugs.launchpad.net/mixxx/+bug/1899242 other than triggering a debug assertion?

**Comment by uklotzde** [comment]
> @uklotzde what are the practical implications of https://bugs.launchpad.net/mixxx/+bug/1899242 other than triggering a debug assertion?

I don't know. Currently no complaints about 2.2 which should use basically the same code.

**Comment by uklotzde** [comment]
Summary: https://bugs.launchpad.net/mixxx/+bug/1914662

**Comment by Be-ing** [comment]
I think there will eventually come a point where we have to rip off the bandage and switch to FFmpeg for everything possible.

**Comment by Be-ing** [comment]
> Currently no complaints about 2.2 which should use basically the same code.

If it isn't broken, maybe we shouldn't rush to fix it for 2.3? I'd be okay with deferring this for 2.4 if the only issue is a debug assertion with unclear practical impact.

**Comment by uklotzde** [comment]
As long as we don't receive bug reports for the release builds we don't need to fix it. And even then the bug won't get fixed without a contributor. We always have the option to switch to FFmpeg if needed!

**Comment by Be-ing** [comment]
Is it possible to fall back to FFMPEG only for files that trigger the debug assertion?

**Comment by uklotzde** [comment]
No, we don't know this in advance. Switching to FFmpeg only for AAC/M4A on Windows would be possible if needed without affecting other platforms and formats.

**Comment by Be-ing** [comment]
If that would be okay to do on Windows, I don't understand why it wouldn't on every OS.

**Comment by uklotzde** [comment]
> If that would be okay to do on Windows, I don't understand why it wouldn't on every OS.

Only as an emergency fallback for 2.3 if needed!

**Comment by uklotzde** [comment]
Closing this now. Disabling DEBUG_ASSERTIONS_FATAL will probably reveal if we need to fix the release version.

**Comment by dan-giddins** [comment]
I am a windows Dev and I am willing to have a look at SoundSourceMediaFoundation.

**Comment by uklotzde** [comment]
> I am a windows Dev and I am willing to have a look at SoundSourceMediaFoundation.

Cool, thank you so much for investigating this issue! If you have any questions about the code, please ask.

We need to find out if those discontinuities are caused by a bookkeeping bug of the decoding position or, if not, how to deal with them. I expect the latter because the file linked in the bug report also doesn't work well with FFmpeg which reports overlapping ranges during decoding. That's probably the reason for the debug assertion.

Since the issues occur during linear playback you don't need to consider seek operations in the first step.

**Comment by dan-giddins** [comment]
@uklotzde Okay, so I am pretty sure it is not a rounding error.

The comments say:
> Both positions should match, otherwise readerFrameIndex is inaccurate due to rounding errors after conversion from stream units to frames! But if this ever happens we better trust m_currentFrameIndex that is continuously updated while reading in forward direction.

However, debugging the code, an example of a failing value is 
```
streamPos = 900005668
m_sampleRate = 44100
kStreamUnitsPerSecond = 10000000
```
With these values, we end up getting to the point where we do a  `round(3969024.99588)`, which is so close to 3969025, its clear that `readerFrameIndex` is mathematically correct, and that double is more than enough precision, however you do the maths. Just to check, I tried changing the order of operations to do `(streamPos *  m_sampleRate) /  kStreamUnitsPerSecond)` instead of the current `(streamPos *  (m_sampleRate /  kStreamUnitsPerSecond)` and that gave me the same answer. The power of double!


However, in this case `m_currentFrameIndex` equals 3969024, so I wonder if something is wrong with `m_currentFrameIndex` instead. The comments around this line assume that there is a rounding error with `readerFrameIndex` and that we should trust `m_currentFrameIndex`, however I don't think we should, and should instead be trusting `readerFrameIndex` (assuming streamPos is correct).

I noticed that to covert from frame index to stream position, we do a `floor()`. In this case `floor()` would make the answer correct, but then it fails in other places, for example `streamPos = 1593360000` leads to a `round(7026717.6)`. If this was a `floor()` it would give us the wrong answer, as it `m_currentFrameIndex = 7026718` in this case. I wonder, is there something contextual that we are missing. What if this does need to be a `floor()` and that it is `m_currentFrameIndex` that is wrong... I will investigate

**Comment by dan-giddins** [comment]
Looking more into this, `m_currentFrameIndex` must also be right, as in this case, each 'frame' is 1024 samples per channel, and 3969024 is a a multiple of 1024. Therefor, `streamPos` must be wrong. The 'ideal' `streamPos` value would be around 900005442 at this point, but we are 226 out from this value. I wonder why.... the mystery continues!

**Comment by dan-giddins** [comment]
@uklotzde is there any reason why we are using `pllTimestamp` over `pdwActualStreamIndex`?
https://docs.microsoft.com/en-us/windows/win32/api/mfreadwrite/nf-mfreadwrite-imfsourcereader-readsample

**Comment by dan-giddins** [comment]
Okay, so this error is definitely coming from `m_pSourceReader->ReadSample()`. `streamPos` (aka `pllTimestamp`) counts up steadily up to this point, but for some reason, suddenly jumps forward by an extra 226... this then causes `readerFrameIndex` to be out by 1.

TLDR: Microsoft Media Foundation source reader is giving us a bad stream position reading. @uklotzde What do we do about this? Do we just rely on `m_currentFrameIndex`?

**Comment by uklotzde** [comment]
Thanks for investigating!!

The rounding issues are expected, unfortunately. I am not sure what to do handle them correctly. Ideally all calculations would be done in integers.

If you only decode the stream linearly without seeking you could simply count samples and ignore the frame position reported by the reader. The complexity starts when you seek to arbitrary positions in the stream and have to resynchronize the decoding positions afterwards. This is required for sample-accurate decoding. You always should read the exact same sample value at any position independent of how you got there. That's the goal.

**Comment by dan-giddins** [comment]
@uklotzde Its not a rounding issue. I've check all of that and we don't lose any precision. I've narrowed it down to the Microsoft Media Foundation source reader sometimes giving us an abnormal stream position.

> The complexity starts when you seek to arbitrary positions in the stream and have to resynchronize the decoding positions afterwards.

Okay I see. I'm really not sure what to do. There must be some kind of bug inside Microsoft Media Foundation source reader and its not open source so I can't fix it!

**Comment by uklotzde** [comment]
I have replaced one of our .m4a test files with the [file linked in the bug report](https://bugs.launchpad.net/mixxx/+bug/1899242/+attachment/5420336/+files/Orjan%20Nilsen%20-%20Endymion%20%28Original%20Mix%29-wKl0dRGfVo4-10s.m4a). The decoding and seek tests work with FAAD, FFmpeg, and CoreAudio (macOS) but fail on Windows when with SoundSource MediaFoundation:

https://github.com/uklotzde/mixxx/runs/1943714313?check_suite_focus=true

Goal: Tests are passing on all platforms for this file.

**Comment by uklotzde** [comment]
I have simply disabled the debug assertion. If the tests pass we might decide that this is good enough and just leave it as is:

https://github.com/uklotzde/mixxx/actions/runs/586286084

**Comment by dan-giddins** [comment]
>I have simply disabled the debug assertion. 

Okay, I think it should be fine. Send me the branch and I could build and use it for a bit to check everything still works fine if you want. Shall we update the comments around the assertion, or just remove it all? 

Just a side note

>Ideally all calculations would be done in integers.

Yes, I agree, but I couldn't seem to find any sample based position marker from the Microsoft Media Foundation source reader. I tried looking at pdwActualStreamIndex, but it just always gives 0. For some reason, only pllTimestamp is available, and this gives us the position by time value and not sample value (hence the conversion). The conversion logic we do is fine, but being given a sample value is not ideal. Even if we where given the sample value, its still very strange how the source reader jumps forward a bit. I would like to investigate this further by looking at `kStreamIndex`, as this seems a bit strange...

**Comment by dan-giddins** [comment]
So `kStreamIndex` isn't relevant here. What is interesting is that the source reader gives us a constant `sampleDuration` for every sample frame of `232200`, but sometimes advances the stream such the the difference between the 'sampleTime' values is significantly different (i.e. enough to give us that mismatch after rounding like `232426`) and sometimes non-significantly different (`232199`). Does anyone know why its being so inconsistent?

**Comment by uklotzde** [comment]
The test in my m4a-test branch is still failing on Windows. The index calculations when the stream discontinuations occur seem to be wrong, which causes differing outputs while decoding the stream.

**Comment by uklotzde** [comment]
`pdwActualStreamIndex` receives the stream index within the container format and is expected to remain constant while decoding. The (audio) stream is selected once when opening the file, see `SoundSourceMediaFoundation::configureAudioStream()`.

**Comment by dan-giddins** [comment]
@uklotzde Okay, I'm really not sure what we can do. We ask the source reader for the next frame of audio, and it sometimes gives us back a `pllTimestamp` time stamp that is too far away from what we were expecting. I've even gone into the buffer and looked at `GetCurrentLength()` and it just gives us back `8192`, even when `pllTimestamp` has jumped...

Just for completeness and for future reference, here is example of the last two frames before the assertion fails (failing on the latter frame):
```
kStreamIndex: 4294967293
streamIndex: 0
dwFlags: 0
streamPos: 899773242
bufferCount: 1
sampleDuration: 232200
sampleFlags: 0
sampleTime: 899773242
totalLength: 8192
bufferLength: 8192

kStreamIndex: 4294967293
streamIndex: 0
dwFlags: 0
streamPos: 900005668
bufferCount: 1
sampleDuration: 232200
sampleFlags: 0
sampleTime: 900005668
totalLength: 8192
bufferLength: 8192
```

**Comment by Be-ing** [comment]
Is Media Foundation just not suitable for our purposes?

**Comment by uklotzde** [comment]
Last try by switching from `round` to `floor` in both directions. We should add comments to https://bugs.launchpad.net/mixxx/+bug/1899242, not here.

**Comment by dan-giddins** [comment]
> Last try by switching from round to floor in both directions. 
@uklotzde I tried that, it breaks it more. There are cases where the `pllTimestamp` is one less than the normal value, which means that the decimal value after the conversion is just below the required value. `floor` rounds it down and breaks it.

**Comment by uklotzde** [comment]
Confirmed. The index calculations are still inconsistent. Maybe broken by design and not fixable. I give up.

**Comment by dan-giddins** [comment]
@uklotzde Interestingly, doing `sampleTime / sampleDuration` gives us a .999 number, implying that if `sampleTime` started at zero, we get a kind of 'drift' over time where sampleTime is a bit less that it should be. I did notice that it was quite often 1 less than  the expected value for time. I suspect there is possibly some kind of loss of precision inside the source reader, and maybe this sudden jump is a way of it 'catching up'? Other than that, I am at a loss.

Is there an open source alternative we can use instead of Media Foundation?

**Comment by Be-ing** [comment]
> Is there an open source alternative we can use instead of Media Foundation?

Yes, FFmpeg as was the original idea with this PR.

**Comment by dan-giddins** [comment]
@Be-ing do you need me to try using this Branch to make sure FFmpeg works on windows? Or have I misunderstood how this hangs together?

**Comment by Be-ing** [comment]
AFAIK there are no issues with FFmpeg itself on Windows. The challenging question is how to deal with the transition away from MediaFoundation. Will this cause users' cue points and beatgrids to shift due to differences in the decoders?

**Comment by dan-giddins** [comment]
@Be-ing As long as we make sure that the output sample rate is the same, sample 'x' should be at the same place, right?

**Comment by uklotzde** [comment]
> @Be-ing As long as we make sure that the output sample rate is the same, sample 'x' should be at the same place, right?

No. Lossy compressed formats are complicated and encoders/decoders behave differently.

**Comment by uklotzde** [comment]
We need to test if and how many files are affected, encoded from different sources.

**Comment by uklotzde** [comment]
But we don't have a choice. Either keep the broken decoder or deal with shifted cue/loop positions and waveforms.

**Comment by daschuer** [comment]
For my understanding the issue is no regression. It is just that is becomes now visible due to fatal debug assertions in our CI builds. Is that correct? 
If so I think it would be reasonable to keep the media foundation  decoder until the offset detection code is in place. This gives us the change to deal with shifted cue points automatically if possible or warn the user that the cue points have been shifted otherwise.

**Comment by Be-ing** [comment]
I am doubtful we will be able to automatically handle the offsets in every case. Nevertheless this is a complex issue that needs quite a bit of work to investigate what files are affected and we should not rush it.

**Comment by daschuer** [comment]
> I am doubtful we will be able to automatically handle the offsets in every case.
Yes. we can only automatically adjust cue point if the offset is caused by an issue in the silence region at the beginning. In all other cases we need the user to readjust the cue points
We cant rush it anyway, because we need the silence detection feature for it which is not available in 2.2
